app/Actions/Photo/Strategies/RotateStrategy.php

```diff
@@ -0,0 +1,182 @@
+		$this->parameters->sourceFileInfo = new SourceFileInfo(
+			pathinfo($oldOriginalFullPath, PATHINFO_BASENAME),
+			$this->photo->type,
+			$tmpFullPath
+		);
+		/** @var SizeVariantNamingStrategy $namingStrategy */
+		$namingStrategy = resolve(SizeVariantNamingStrategy::class);
+		$namingStrategy->setFallbackExtension($this->parameters->sourceFileInfo->getOriginalFileExtension());
+		/** @var SizeVariantFactory $sizeVariantFactory */
+		$sizeVariantFactory = resolve(SizeVariantFactory::class);
+		$sizeVariantFactory->init($this->photo, $namingStrategy);
+
+		// Create size variant for rotated original
+		// Note that his also creates a different file name than before
+		// because the checksum of the photo has changed.
+		// Using a different filename allows to avoid caching effects.
+		// Sic! Swap width and height here, because the image has been rotated
+		$newOriginalSizeVariant = $sizeVariantFactory->createOriginal($oldOriginalHeight, $oldOriginalWidth);
+		$this->putSourceIntoFinalDestination($newOriginalSizeVariant->full_path);
+
+		// Create remaining size variants
+		$newSizeVariants = null;
+		try {
+			$newSizeVariants = $sizeVariantFactory->createSizeVariants();
+			// add new original size variant to collection of newly created
+			// size variant; we need this to correctly update the duplicates
+			// below
+			$newSizeVariants->add($newOriginalSizeVariant);
+		} catch (\Throwable $t) {
+			// Don't re-throw the exception, because we do not want the
+			// import to fail completely only due to missing size variants.
+			// There are just too many options why the creation of size
+			// variants may fail: the user has uploaded an unsupported file
+			// format, GD and Imagick are both not available or disabled
+			// by configuration, etc.
+			Logs::error(__METHOD__, __LINE__, 'Failed to generate size variants, error was ' . $t->getMessage());
+		}
+
+		// Clean up factory
+		$sizeVariantFactory->cleanup();
+
+		// Deal with duplicates.  We simply update all of them to match.
+		$duplicates = Photo::query()
+			->where('checksum', $oldChecksum)
+			->get();
+		/** @var Photo $duplicate */
+		foreach ($duplicates as $duplicate) {
+			$duplicate->filesize = $this->photo->filesize;
+			$duplicate->checksum = $this->photo->checksum;
+			// Note: It is not correct to simply update the existing size
+			// variants of the duplicates.
+			// Due to rotation the number and type of size variants may have
+			// changed, too.
+			// So we actually have to do a 3-way merge and update:
+			// a) delete size variants of the duplicates which do not exist
+			// anymore, b) update size variants of the duplicates which
+			// still exist and c) add new size variants to duplicates which
+			// haven't existed before.
+			// For simplicity, we simply delete all size variants of the
+			// duplicates and re-create them.
+			// Deleting the size variants of the duplicates has also the
+			// advantage that the actual files are erased from storage.
```

Comment:
Out of curiosity, when do the actual files get deleted if a photo had duplicates? In line 100 above (`$this->photo->size_variants->delete();`) or when the (last) duplicate is being updated here in this loop?

Also, so each duplicate has its own copy of `size_variants`? That's too bad; if they all shared one set, we wouldn't need this code below...

Reply:
> Out of curiosity, when do the actual files get deleted if a photo had duplicates? In line 100 above (`$this->photo->size_variants->delete();`) or when the (last) duplicate is being updated here in this loop?

The old files get deleted when the (last) duplicate is being updated.

> Also, so each duplicate has its own copy of `size_variants`? That's too bad; if they all shared one set, we wouldn't need this code below...

I know. I already proposed such an idea around one year ago even before I had started coding on Lychee. However, not the photos should share the same size variants, but albums should share the same photo. (That was my idea back then.) So what I did propose was to have an m:n-relation between albums and photos and an intermediate table `AlbumPhotoAssociation` in between. I still believe that this would be the right approach and would allow us to drop some code. However @ildyria was not a fan of having a m:n relationship. (I still don't know why.) But this would be (if ever) another PR and needs some discussion on Gitter first.